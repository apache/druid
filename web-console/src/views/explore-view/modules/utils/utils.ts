/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import type { SqlExpression } from '@druid-toolkit/query';
import { F, SqlFunction, SqlLiteral } from '@druid-toolkit/query';

export function shiftTimeInWhere(where: SqlExpression, period: string): SqlExpression {
  return where.walk(ex => {
    if (ex instanceof SqlLiteral) {
      // Works with: __time < TIMESTAMP '2022-01-02 03:04:05'
      if (ex.isDate()) {
        return F('TIME_SHIFT', ex, period, -1);
      }
    } else if (ex instanceof SqlFunction) {
      const effectiveFunctionName = ex.getEffectiveFunctionName();

      // Works with: TIME_IN_INTERVAL(__time, '<interval>')
      if (effectiveFunctionName === 'TIME_IN_INTERVAL') {
        return ex.changeArgs(ex.args!.change(0, F('TIME_SHIFT', ex.getArg(0), period, 1)));
      }

      // Works with: TIME_SHIFT(...) <= __time
      //        and: __time < MAX_DATA_TIME()
      if (effectiveFunctionName === 'TIME_SHIFT' || effectiveFunctionName === 'MAX_DATA_TIME') {
        return F('TIME_SHIFT', ex, period, -1);
      }
    }

    return ex;
  }) as SqlExpression;
}
