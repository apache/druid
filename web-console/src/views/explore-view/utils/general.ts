/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import type { SqlExpression } from '@druid-toolkit/query';
import { SqlColumn } from '@druid-toolkit/query';

export function addTableScope(expression: SqlExpression, newTableScope: string): SqlExpression {
  return expression.walk(ex => {
    if (ex instanceof SqlColumn && !ex.getTableName()) {
      return ex.changeTableName(newTableScope);
    }
    return ex;
  }) as SqlExpression;
}

export type Rename = Map<string, string>;

export function renameColumnsInExpression(
  expression: SqlExpression,
  rename: Rename,
): SqlExpression {
  return expression.walk(ex => {
    if (ex instanceof SqlColumn) {
      const renameTo = rename.get(ex.getName());
      if (renameTo) {
        return ex.changeName(renameTo);
      }
    }
    return ex;
  }) as SqlExpression;
}

export function changeOrAdd<T>(xs: readonly T[], oldValue: T | undefined, newValue: T): T[] {
  if (typeof oldValue === 'undefined') {
    return xs.concat([newValue]);
  } else {
    return xs.map(x => (x === oldValue ? newValue : x));
  }
}
