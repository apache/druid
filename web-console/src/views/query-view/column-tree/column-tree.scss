/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

@import '../../../blueprint-overrides/common/colors';

.column-tree {
  .bp3-dark & {
    background: $dark-gray3;
  }

  .schema-selector {
    position: absolute;
    top: 0;

    select {
      border-bottom-right-radius: 0;
      border-bottom-left-radius: 0;
    }
  }

  .tree-container {
    position: absolute;
    top: 40px;
    bottom: 0;
    width: 100%;
    overflow: auto;

    .bp3-tree-node-content-1 {
      padding-left: 0;
    }
  }

  .bp3-popover-wrapper {
    width: 188px;
    display: inline-block;

    .bp3-popover-target {
      width: 100%;
      display: inline-block;
      cursor: pointer;
    }
  }
}
