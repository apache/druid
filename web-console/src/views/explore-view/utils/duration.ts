/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { sum } from 'd3-array';

import { filterMap, pluralIfNeeded } from '../../../utils';

const SPANS = ['year', 'month', 'day', 'hour', 'minute', 'second'];

export function formatDuration(duration: string, preferCompact = false): string {
  // Regular expressions to match ISO 8601 duration parts
  const regex = /P(?:(\d+)Y)?(?:(\d+)M)?(?:(\d+)D)?(?:T(?:(\d+)H)?(?:(\d+)M)?(?:(\d+)S)?)?/;
  const matches = duration.match(regex);
  if (!matches) return duration;

  // Extract the relevant parts
  const counts = SPANS.map((_, i) => parseInt(matches[i + 1] || '0', 10));

  // Construct the human-readable format based on the parsed values
  let parts: string[];
  if (preferCompact && sum(counts) === 1) {
    parts = filterMap(SPANS, (span, i) => (counts[i] ? span : undefined));
  } else {
    parts = filterMap(SPANS, (span, i) =>
      counts[i] ? pluralIfNeeded(counts[i], span) : undefined,
    );
  }

  // Join the parts with commas, and return the result
  return parts.length > 0 ? parts.join(', ') : '0 seconds';
}
