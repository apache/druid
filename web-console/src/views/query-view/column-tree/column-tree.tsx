/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { HTMLSelect, ITreeNode, Menu, MenuItem, Popover, Position, Tree } from '@blueprintjs/core';
import { IconNames } from '@blueprintjs/icons';
import {
  SqlAlias,
  SqlComparison,
  SqlExpression,
  SqlFunction,
  SqlJoinPart,
  SqlQuery,
  SqlRef,
} from 'druid-query-toolkit';
import React, { ChangeEvent } from 'react';

import { Loader } from '../../../components';
import { Deferred } from '../../../components/deferred/deferred';
import { ColumnMetadata, copyAndAlert, groupBy, oneOf, prettyPrintSql } from '../../../utils';
import { dataTypeToIcon } from '../query-utils';

import { NumberMenuItems, StringMenuItems, TimeMenuItems } from './column-tree-menu';

import './column-tree.scss';

const LAST_DAY = SqlExpression.parse(`__time >= CURRENT_TIMESTAMP - INTERVAL '1' DAY`);
const COUNT_STAR = SqlFunction.COUNT_STAR.as('Count');

const STRING_QUERY = SqlQuery.parse(`SELECT
  ?
FROM ?
GROUP BY 1
ORDER BY 2 DESC`);

const TIME_QUERY = SqlQuery.parse(`SELECT
  TIME_FLOOR(?, 'PT1H') AS "Time"
FROM ?
GROUP BY 1
ORDER BY 1 ASC`);

interface HandleColumnClickOptions {
  columnSchema: string;
  columnTable: string;
  columnName: string;
  columnType: string;
  parsedQuery: SqlQuery | undefined;
  onQueryChange: (query: SqlQuery, run: boolean) => void;
}

function handleColumnClick(options: HandleColumnClickOptions): void {
  const { columnSchema, columnTable, columnName, columnType, parsedQuery, onQueryChange } = options;

  let query: SqlQuery;
  const columnRef = SqlRef.column(columnName);
  if (columnSchema === 'druid') {
    if (columnType === 'TIMESTAMP') {
      query = TIME_QUERY.fillPlaceholders([columnRef, SqlRef.table(columnTable)]) as SqlQuery;
    } else {
      query = STRING_QUERY.fillPlaceholders([columnRef, SqlRef.table(columnTable)]) as SqlQuery;
    }
  } else {
    query = STRING_QUERY.fillPlaceholders([
      columnRef,
      SqlRef.table(columnTable, columnSchema),
    ]) as SqlQuery;
  }

  let where: SqlExpression | undefined;
  let aggregates: SqlAlias[] = [];
  if (parsedQuery && parsedQuery.getFirstTableName() === columnTable) {
    where = parsedQuery.getWhereExpression();
    aggregates = parsedQuery.getAggregateSelectExpressions();
  } else if (columnSchema === 'druid') {
    where = LAST_DAY;
  }
  if (!aggregates.length) {
    aggregates.push(COUNT_STAR);
  }

  let newSelectExpressions = query.selectExpressions;
  for (const aggregate of aggregates) {
    newSelectExpressions = newSelectExpressions.append(aggregate);
  }

  onQueryChange(
    query.changeSelectExpressions(newSelectExpressions).changeWhereExpression(where),
    true,
  );
}

export interface ColumnTreeProps {
  columnMetadataLoading: boolean;
  columnMetadata?: readonly ColumnMetadata[];
  getParsedQuery: () => SqlQuery | undefined;
  onQueryChange: (query: SqlQuery, run?: boolean) => void;
  defaultSchema?: string;
  defaultTable?: string;
}

export interface ColumnTreeState {
  prevColumnMetadata?: readonly ColumnMetadata[];
  columnTree?: ITreeNode[];
  currentSchemaSubtree?: ITreeNode[];
  selectedTreeIndex: number;
}

export function getJoinColumns(parsedQuery: SqlQuery, _table: string) {
  let lookupColumn: string | undefined;
  let originalTableColumn: string | undefined;
  if (parsedQuery.fromClause && parsedQuery.fromClause.joinParts) {
    const firstOnExpression = parsedQuery.fromClause.joinParts.first().onExpression;
    if (firstOnExpression instanceof SqlComparison && firstOnExpression.op === '=') {
      const { lhs, rhs } = firstOnExpression;
      if (lhs instanceof SqlRef && lhs.namespace === 'lookup') {
        lookupColumn = lhs.column;
      }
      if (rhs instanceof SqlRef) {
        originalTableColumn = rhs.column;
      }
    }
  }

  return {
    lookupColumn: lookupColumn || 'k',
    originalTableColumn: originalTableColumn || 'XXX',
  };
}

export class ColumnTree extends React.PureComponent<ColumnTreeProps, ColumnTreeState> {
  static getDerivedStateFromProps(props: ColumnTreeProps, state: ColumnTreeState) {
    const { columnMetadata, defaultSchema, defaultTable, onQueryChange } = props;

    if (columnMetadata && columnMetadata !== state.prevColumnMetadata) {
      const columnTree = groupBy(
        columnMetadata,
        r => r.TABLE_SCHEMA,
        (metadata, schemaName): ITreeNode => ({
          id: schemaName,
          label: schemaName,
          childNodes: groupBy(
            metadata,
            r => r.TABLE_NAME,
            (metadata, tableName): ITreeNode => ({
              id: tableName,
              icon: IconNames.TH,
              label: (
                <Popover
                  boundary={'window'}
                  position={Position.RIGHT}
                  content={
                    <Deferred
                      content={() => {
                        const parsedQuery = props.getParsedQuery();
                        const tableRef = SqlRef.table(tableName).as();
                        const prettyTableRef = prettyPrintSql(tableRef);
                        return (
                          <Menu>
                            <MenuItem
                              icon={IconNames.FULLSCREEN}
                              text={`SELECT ... FROM ${tableName}`}
                              onClick={() => {
                                const tableRef = SqlRef.table(
                                  tableName,
                                  schemaName === 'druid' ? undefined : schemaName,
                                );

                                let where: SqlExpression | undefined;
                                if (parsedQuery && parsedQuery.getFirstTableName() === tableName) {
                                  where = parsedQuery.getWhereExpression();
                                } else if (schemaName === 'druid') {
                                  where = LAST_DAY;
                                }

                                onQueryChange(
                                  SqlQuery.create(tableRef)
                                    .changeSelectExpressions(
                                      metadata.map(child => SqlRef.column(child.COLUMN_NAME).as()),
                                    )
                                    .changeWhereExpression(where),
                                  true,
                                );
                              }}
                            />
                            {parsedQuery && parsedQuery.getFirstTableName() !== tableName && (
                              <MenuItem
                                icon={IconNames.EXCHANGE}
                                text={`Replace FROM with: ${prettyTableRef}`}
                                onClick={() => {
                                  onQueryChange(
                                    parsedQuery.changeFromExpressions([tableRef]),
                                    true,
                                  );
                                }}
                              />
                            )}
                            {parsedQuery && schemaName === 'lookup' && (
                              <MenuItem
                                popoverProps={{ openOnTargetFocus: false }}
                                icon={IconNames.JOIN_TABLE}
                                text={parsedQuery.hasJoin() ? `Replace join` : `Join`}
                              >
                                <MenuItem
                                  icon={IconNames.LEFT_JOIN}
                                  text={`Left join`}
                                  onClick={() => {
                                    const { lookupColumn, originalTableColumn } = getJoinColumns(
                                      parsedQuery,
                                      tableName,
                                    );
                                    onQueryChange(
                                      parsedQuery
                                        .removeAllJoins()
                                        .addJoin(
                                          SqlJoinPart.create(
                                            'LEFT',
                                            SqlRef.column(tableName, schemaName).upgrade(),
                                            SqlRef.column(lookupColumn, tableName, 'lookup').equal(
                                              SqlRef.column(
                                                originalTableColumn,
                                                parsedQuery.getFirstTableName(),
                                              ),
                                            ),
                                          ),
                                        ),
                                      false,
                                    );
                                  }}
                                />
                                <MenuItem
                                  icon={IconNames.INNER_JOIN}
                                  text={`Inner join`}
                                  onClick={() => {
                                    const { lookupColumn, originalTableColumn } = getJoinColumns(
                                      parsedQuery,
                                      tableName,
                                    );
                                    onQueryChange(
                                      parsedQuery.addJoin(
                                        SqlJoinPart.create(
                                          'INNER',
                                          SqlRef.column(tableName, schemaName).upgrade(),
                                          SqlRef.column(lookupColumn, tableName, 'lookup').equal(
                                            SqlRef.column(
                                              originalTableColumn,
                                              parsedQuery.getFirstTableName(),
                                            ),
                                          ),
                                        ),
                                      ),
                                      false,
                                    );
                                  }}
                                />
                              </MenuItem>
                            )}
                            {parsedQuery &&
                              parsedQuery.hasJoin() &&
                              parsedQuery.getJoins()[0].table.toString() === tableName && (
                                <MenuItem
                                  icon={IconNames.EXCHANGE}
                                  text={`Remove join`}
                                  onClick={() => onQueryChange(parsedQuery.removeAllJoins())}
                                />
                              )}
                            {parsedQuery &&
                              parsedQuery.hasGroupBy() &&
                              parsedQuery.getFirstTableName() === tableName && (
                                <MenuItem
                                  icon={IconNames.FUNCTION}
                                  text={`Aggregate COUNT(*)`}
                                  onClick={() =>
                                    onQueryChange(parsedQuery.addSelectExpression(COUNT_STAR), true)
                                  }
                                />
                              )}
                            <MenuItem
                              icon={IconNames.CLIPBOARD}
                              text={`Copy: ${prettyTableRef}`}
                              onClick={() => {
                                copyAndAlert(
                                  tableRef.toString(),
                                  `${prettyTableRef} query copied to clipboard`,
                                );
                              }}
                            />
                          </Menu>
                        );
                      }}
                    />
                  }
                >
                  {tableName}
                </Popover>
              ),
              childNodes: metadata
                .map(
                  (columnData): ITreeNode => ({
                    id: columnData.COLUMN_NAME,
                    icon: dataTypeToIcon(columnData.DATA_TYPE),
                    label: (
                      <Popover
                        boundary={'window'}
                        position={Position.RIGHT}
                        autoFocus={false}
                        targetClassName={'bp3-popover-open'}
                        content={
                          <Deferred
                            content={() => {
                              const parsedQuery = props.getParsedQuery();
                              return (
                                <Menu>
                                  <MenuItem
                                    icon={IconNames.FULLSCREEN}
                                    text={`Show: ${columnData.COLUMN_NAME}`}
                                    onClick={() => {
                                      handleColumnClick({
                                        columnSchema: schemaName,
                                        columnTable: tableName,
                                        columnName: columnData.COLUMN_NAME,
                                        columnType: columnData.DATA_TYPE,
                                        parsedQuery,
                                        onQueryChange: onQueryChange,
                                      });
                                    }}
                                  />
                                  {parsedQuery &&
                                    oneOf(columnData.DATA_TYPE, 'BIGINT', 'FLOAT', 'DOUBLE') && (
                                      <NumberMenuItems
                                        table={tableName}
                                        schema={schemaName}
                                        columnName={columnData.COLUMN_NAME}
                                        parsedQuery={parsedQuery}
                                        onQueryChange={onQueryChange}
                                      />
                                    )}
                                  {parsedQuery && columnData.DATA_TYPE === 'VARCHAR' && (
                                    <StringMenuItems
                                      table={tableName}
                                      schema={schemaName}
                                      columnName={columnData.COLUMN_NAME}
                                      parsedQuery={parsedQuery}
                                      onQueryChange={onQueryChange}
                                    />
                                  )}
                                  {parsedQuery && columnData.DATA_TYPE === 'TIMESTAMP' && (
                                    <TimeMenuItems
                                      table={tableName}
                                      schema={schemaName}
                                      columnName={columnData.COLUMN_NAME}
                                      parsedQuery={parsedQuery}
                                      onQueryChange={onQueryChange}
                                    />
                                  )}
                                  <MenuItem
                                    icon={IconNames.CLIPBOARD}
                                    text={`Copy: ${columnData.COLUMN_NAME}`}
                                    onClick={() => {
                                      copyAndAlert(
                                        columnData.COLUMN_NAME,
                                        `${columnData.COLUMN_NAME} query copied to clipboard`,
                                      );
                                    }}
                                  />
                                </Menu>
                              );
                            }}
                          />
                        }
                      >
                        {columnData.COLUMN_NAME}
                      </Popover>
                    ),
                  }),
                )
                .sort((a, b) =>
                  String(a.id)
                    .toLowerCase()
                    .localeCompare(String(b.id).toLowerCase()),
                ),
            }),
          ),
        }),
      );

      let selectedTreeIndex = -1;
      let expandedNode = -1;
      if (defaultSchema && columnTree) {
        selectedTreeIndex = columnTree.findIndex(x => {
          return x.id === defaultSchema;
        });
      }

      if (selectedTreeIndex > -1) {
        const treeNodes = columnTree[selectedTreeIndex].childNodes;
        if (treeNodes) {
          if (defaultTable) {
            expandedNode = treeNodes.findIndex(node => {
              return node.id === defaultTable;
            });
          }
        }
      }

      if (!columnTree) return null;
      const currentSchemaSubtree =
        columnTree[selectedTreeIndex > -1 ? selectedTreeIndex : 0].childNodes;
      if (!currentSchemaSubtree) return null;

      if (expandedNode > -1) {
        currentSchemaSubtree[expandedNode].isExpanded = true;
      }

      return {
        prevColumnMetadata: columnMetadata,
        columnTree,
        selectedTreeIndex,
        currentSchemaSubtree,
      };
    }
    return null;
  }

  constructor(props: ColumnTreeProps, context: any) {
    super(props, context);
    this.state = {
      selectedTreeIndex: -1,
    };
  }

  renderSchemaSelector() {
    const { columnTree, selectedTreeIndex } = this.state;
    if (!columnTree) return null;

    return (
      <HTMLSelect
        className="schema-selector"
        value={selectedTreeIndex > -1 ? selectedTreeIndex : undefined}
        onChange={this.handleSchemaSelectorChange}
        fill
        minimal
        large
      >
        {columnTree.map((treeNode, i) => (
          <option key={i} value={i}>
            {treeNode.label}
          </option>
        ))}
      </HTMLSelect>
    );
  }

  private handleSchemaSelectorChange = (e: ChangeEvent<HTMLSelectElement>): void => {
    const { columnTree } = this.state;

    const selectedTreeIndex = Number(e.target.value);

    if (!columnTree) return;

    const currentSchemaSubtree =
      columnTree[selectedTreeIndex > -1 ? selectedTreeIndex : 0].childNodes;

    this.setState({
      selectedTreeIndex: Number(e.target.value),
      currentSchemaSubtree: currentSchemaSubtree,
    });
  };

  private handleNodeCollapse = (nodeData: ITreeNode) => {
    nodeData.isExpanded = false;
    this.bounceState();
  };

  private handleNodeExpand = (nodeData: ITreeNode) => {
    nodeData.isExpanded = true;
    this.bounceState();
  };

  bounceState() {
    const { columnTree } = this.state;
    if (!columnTree) return;
    this.setState(prevState => ({
      columnTree: prevState.columnTree ? prevState.columnTree.slice() : undefined,
    }));
  }

  render(): JSX.Element | null {
    const { columnMetadataLoading } = this.props;
    const { currentSchemaSubtree } = this.state;

    if (columnMetadataLoading) {
      return (
        <div className="column-tree">
          <Loader />
        </div>
      );
    }

    if (!currentSchemaSubtree) return null;

    return (
      <div className="column-tree">
        {this.renderSchemaSelector()}
        <div className="tree-container">
          <Tree
            contents={currentSchemaSubtree}
            onNodeCollapse={this.handleNodeCollapse}
            onNodeExpand={this.handleNodeExpand}
          />
        </div>
      </div>
    );
  }
}
