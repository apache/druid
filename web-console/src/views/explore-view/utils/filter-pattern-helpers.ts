/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import type { Column, FilterPattern } from '@druid-toolkit/query';

import { DATE_FORMAT } from './date-format';
import { formatDuration } from './duration';

const TIME_RELATIVE_TYPES: Record<string, string> = {
  'maxDataTime/': 'latest',
  'timestamp/ceil': 'current',
  'timestamp/floor': 'previous',
};

export function initPatternForColumn(column: Column): FilterPattern {
  switch (column.sqlType) {
    case 'TIMESTAMP':
      return {
        type: 'timeRelative',
        negated: false,
        column: column.name,
        anchor: 'timestamp',
        rangeDuration: 'P1D',
        startBound: '[',
        endBound: ')',
      };

    case 'BIGINT':
    case 'FLOAT':
    case 'DOUBLE':
      return {
        type: 'numberRange',
        negated: false,
        column: column.name,
        start: 0,
        startBound: '[',
        end: 100,
        endBound: ')',
      };

    default:
      return {
        type: 'values',
        negated: false,
        column: column.name,
        values: [],
      };
  }
}

export function formatPatternWithoutNegation(pattern: FilterPattern): string {
  switch (pattern.type) {
    case 'values':
      return `${pattern.column}: ${pattern.values
        .map(v => (v === '' ? 'empty' : String(v)))
        .join(', ')}`;

    case 'contains':
      return `${pattern.column} ~ '${pattern.contains}'`;

    case 'regexp':
      return `${pattern.column} ~ /${pattern.regexp}/`;

    case 'timeInterval': {
      return DATE_FORMAT.formatRange(pattern.start, pattern.end);
    }

    case 'timeRelative': {
      const type = TIME_RELATIVE_TYPES[`${pattern.anchor}/${pattern.alignType || ''}`];
      return `${pattern.column} in ${type ? `${type} ` : ''}${formatDuration(
        pattern.rangeDuration,
        true,
      )}`;
    }

    case 'numberRange':
      return `${pattern.column} in ${pattern.startBound}${pattern.start}, ${pattern.end}${pattern.endBound}`;

    case 'mvContains':
      return `${pattern.column} on of ${pattern.values
        .map(v => (v === '' ? 'empty' : String(v)))
        .join(', ')}`;

    case 'custom':
      return String(pattern.expression);
  }
}
