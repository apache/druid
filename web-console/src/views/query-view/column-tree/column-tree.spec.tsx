/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { SqlQuery } from 'druid-query-toolkit';
import { shallow } from 'enzyme';
import React from 'react';

import { ColumnMetadata } from '../../../utils';

import { ColumnTree } from './column-tree';

describe('ColumnTree', () => {
  it('matches snapshot', () => {
    const columnTree = shallow(
      <ColumnTree
        getParsedQuery={() => {
          return SqlQuery.parse(`SELECT channel, count(*) as cnt FROM wikipedia GROUP BY 1`);
        }}
        defaultSchema="druid"
        defaultTable="wikipedia"
        columnMetadataLoading={false}
        columnMetadata={
          [
            {
              TABLE_SCHEMA: 'druid',
              TABLE_NAME: 'wikipedia',
              COLUMN_NAME: '__time',
              DATA_TYPE: 'TIMESTAMP',
            },
            {
              TABLE_SCHEMA: 'druid',
              TABLE_NAME: 'wikipedia',
              COLUMN_NAME: 'added',
              DATA_TYPE: 'BIGINT',
            },
            {
              TABLE_SCHEMA: 'druid',
              TABLE_NAME: 'wikipedia',
              COLUMN_NAME: 'addedBy10',
              DATA_TYPE: 'FLOAT',
            },
            {
              TABLE_SCHEMA: 'sys',
              TABLE_NAME: 'tasks',
              COLUMN_NAME: 'error_msg',
              DATA_TYPE: 'VARCHAR',
            },
          ] as ColumnMetadata[]
        }
        onQueryChange={() => {}}
      />,
    );

    expect(columnTree).toMatchSnapshot();
  });
});
