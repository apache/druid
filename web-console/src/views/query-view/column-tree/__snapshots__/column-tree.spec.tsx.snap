// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ColumnTree matches snapshot 1`] = `
<div
  className="column-tree"
>
  <HTMLSelect
    className="schema-selector"
    fill={true}
    large={true}
    minimal={true}
    onChange={[Function]}
    value={0}
  >
    <option
      key="0"
      value={0}
    >
      druid
    </option>
    <option
      key="1"
      value={1}
    >
      sys
    </option>
  </HTMLSelect>
  <div
    className="tree-container"
  >
    <Blueprint3.Tree
      contents={
        Array [
          Object {
            "childNodes": Array [
              Object {
                "icon": "time",
                "id": "__time",
                "label": <Blueprint3.Popover
                  autoFocus={false}
                  boundary="window"
                  captureDismiss={false}
                  content={
                    <Memo(Deferred)
                      content={[Function]}
                    />
                  }
                  defaultIsOpen={false}
                  disabled={false}
                  fill={false}
                  hasBackdrop={false}
                  hoverCloseDelay={300}
                  hoverOpenDelay={150}
                  inheritDarkTheme={true}
                  interactionKind="click"
                  minimal={false}
                  modifiers={Object {}}
                  openOnTargetFocus={true}
                  position="right"
                  targetClassName="bp3-popover-open"
                  targetTagName="span"
                  transitionDuration={300}
                  usePortal={true}
                  wrapperTagName="span"
                >
                  __time
                </Blueprint3.Popover>,
              },
              Object {
                "icon": "numerical",
                "id": "added",
                "label": <Blueprint3.Popover
                  autoFocus={false}
                  boundary="window"
                  captureDismiss={false}
                  content={
                    <Memo(Deferred)
                      content={[Function]}
                    />
                  }
                  defaultIsOpen={false}
                  disabled={false}
                  fill={false}
                  hasBackdrop={false}
                  hoverCloseDelay={300}
                  hoverOpenDelay={150}
                  inheritDarkTheme={true}
                  interactionKind="click"
                  minimal={false}
                  modifiers={Object {}}
                  openOnTargetFocus={true}
                  position="right"
                  targetClassName="bp3-popover-open"
                  targetTagName="span"
                  transitionDuration={300}
                  usePortal={true}
                  wrapperTagName="span"
                >
                  added
                </Blueprint3.Popover>,
              },
              Object {
                "icon": "numerical",
                "id": "addedBy10",
                "label": <Blueprint3.Popover
                  autoFocus={false}
                  boundary="window"
                  captureDismiss={false}
                  content={
                    <Memo(Deferred)
                      content={[Function]}
                    />
                  }
                  defaultIsOpen={false}
                  disabled={false}
                  fill={false}
                  hasBackdrop={false}
                  hoverCloseDelay={300}
                  hoverOpenDelay={150}
                  inheritDarkTheme={true}
                  interactionKind="click"
                  minimal={false}
                  modifiers={Object {}}
                  openOnTargetFocus={true}
                  position="right"
                  targetClassName="bp3-popover-open"
                  targetTagName="span"
                  transitionDuration={300}
                  usePortal={true}
                  wrapperTagName="span"
                >
                  addedBy10
                </Blueprint3.Popover>,
              },
            ],
            "icon": "th",
            "id": "wikipedia",
            "isExpanded": true,
            "label": <Blueprint3.Popover
              boundary="window"
              captureDismiss={false}
              content={
                <Memo(Deferred)
                  content={[Function]}
                />
              }
              defaultIsOpen={false}
              disabled={false}
              fill={false}
              hasBackdrop={false}
              hoverCloseDelay={300}
              hoverOpenDelay={150}
              inheritDarkTheme={true}
              interactionKind="click"
              minimal={false}
              modifiers={Object {}}
              openOnTargetFocus={true}
              position="right"
              targetTagName="span"
              transitionDuration={300}
              usePortal={true}
              wrapperTagName="span"
            >
              wikipedia
            </Blueprint3.Popover>,
          },
        ]
      }
      onNodeCollapse={[Function]}
      onNodeExpand={[Function]}
    />
  </div>
</div>
`;
